Add utility classes for empirical and probability distributions

cqc-alec opened the pull request:
@rossduncan I am adding you as a reviewer since this was mainly your request. Please let me know if you think there is essential functionality I have missed (sampling? Shannon entropy?... this sort of thing can be added if needed) -- or if I have struck the wrong level of abstraction.

Distributions can be derived from `BackendResult` objects, but you can't go back the other way. This didn't make much sense to me, since they are essentially more abstract structures and can be manipulated in ways that make reversal meaningless. Was there a concrete use case for this?

ss2165 reviewed (commented):
I will need to take a more detailed look to check correctness etc. but I thought I'd leave a couple of broad comments first:

1. I like the idea and the interfaces, very neat.
2. I wonder though if we should be defining all the statistical functions ourselves. Perhaps we can [use a scipy discrete distribution](https://docs.scipy.org/doc/scipy/tutorial/stats.html#subclassing-rv-discrete) for `ProbabilityDistribution` and then just pair that with the total number of samples to get `EmpricalDistribution` (which, incidentally, are called `Distribution` and `SampleDistribution` in Tierkreis-land)

cqc-alec commented:
>     2. I wonder though if we should be defining all the statistical functions ourselves. Perhaps we can [use a scipy discrete distribution](https://docs.scipy.org/doc/scipy/tutorial/stats.html#subclassing-rv-discrete) for `ProbabilityDistribution` and then just pair that with the total number of samples to get `EmpricalDistribution` (which, incidentally, are called `Distribution` and `SampleDistribution` in Tierkreis-land)

It seems that `scipy.stats.rv_discrete` is limited to integer-valued distributions (the distributions we deal with mostly aren't "naturally" integer-valued), and also to distributions with full support (I wanted to include partial distributions -- for one thing this makes convex combinations easy to express). So I don't think it quite meets our needs. (I was a bit surprised not to find an existing library that does though.)

I think I prefer not to make an `EmpiricalDistribution` just a `ProbabilityDistribution` with a sample size, even if in principle it does capture the same information (modulo rounding errors, which could be fiddly). Some operations make sense for one and not the other, so I don't think it would actually save much code, and could be more bug-prone.

Happy to rename to `Distribution` and `SampleDistribution` if that is preferred.

cqc-alec commented:
Following offline discussions, I have changed `ProbabilityDistribution` to be normalized-only, and added a separate `convex_combination()` method. I have also added a method to convert a `ProbabilityDistribution` to a `scipy.stats.rv_discrete` object together with a list showing the correspondence of values (since `rv_discrete` RVs are integer-valued).

ss2165 commented on pytket/pytket/utils/distribution.py:
```diff
@@ -0,0 +1,268 @@
+# limitations under the License.
+
+from collections import Counter, defaultdict
+from typing import cast, Any, Callable, DefaultDict, Dict, List, Set, Tuple, Union
```
unused cast

cqc-alec replied:
> the only major thing is I think it's worth using generic typing for this, I've demonstrated it here: #653 but I'm not actually sure if mypy is happy because I can't get it to run on my machine (my IDE type checking is happy at least)

That looks great, I'll check locally. I didn't know about `typing.Generic` before.

What do `OT` and `NT` stand for?

cqc-alec replied:
> unused cast

Surprised pylint didn't pick this up...

cqc-alec replied:
Fixed (by your commit)..

ss2165 replied:
> What do `OT` and `NT` stand for?

"Observable Type" and "New Type" I think, was just matching the examples in the generic docs

ss2165 commented on pytket/pytket/utils/distribution.py:
```diff
@@ -0,0 +1,268 @@
+    def __init__(self, P: Dict[Any, float]):
+        """Initialize with a dictionary of probabilities.
+
+        The values must be non-negative and add up to at most 1.
```
```suggestion
        The values must be non-negative and add up to 1.
```

cqc-alec replied:
Good spot.

cqc-alec replied:
Fixed,

ss2165 commented on pytket/pytket/utils/distribution.py:
```diff
@@ -0,0 +1,268 @@
+        """
+        if any(x < 0 for x in P.values()):
+            raise ValueError("Distribution contains negative probabilities")
+        S = sum(P.values())
+        if not np.isclose(S, 1):
+            raise ValueError("Probabilities do not sum to 1")
+        self._P: Dict[Any, float] = {x: p for x, p in P.items() if not np.isclose(p, 0)}
```
Could the `isclose` to 1 check for the sum fail if all the `isclose` to zero values are removed?

cqc-alec replied:
In theory, yes. We could allow construction from non-normalized values (with scaling applied in the constructor), to get around this.

cqc-alec replied:
Fixed,

ss2165 commented on pytket/tests/distribution_test.py:
```diff
@@ -0,0 +1,105 @@
+    pd4 = convex_combination([(pd, 1 / 3), (pd3, 2 / 3)])
+    assert pd4 == ProbabilityDistribution({"a": 7 / 27, "b": 14 / 27, "c": 2 / 9})
+
+    ed1 = EmpiricalDistribution(Counter({"a": 1, "b": 2}))
```
maybe this line onwards should be a separate test?

cqc-alec replied:
Done.

ss2165 reviewed (changes requested):
Looks good just a couple of minor things here

the only major thing is I think it's worth using generic typing for this, I've demonstrated it here: https://github.com/CQCL/tket/pull/653 but I'm not actually sure if mypy is happy because I can't get it to run on my machine (my IDE type checking is happy at least)

cqc-alec commented on pytket/pytket/utils/distribution.py:
```diff
@@ -0,0 +1,286 @@
+        """Return the distribution as a :py:class:`dict` object."""
+        return self._P
+
+    def as_rv_discrete(self) -> Tuple[rv_discrete, List]:
```
Thanks, I added a couple more I spotted.

cqc-alec commented on pytket/pytket/utils/distribution.py:
```diff
@@ -0,0 +1,286 @@
+        S = sum(p for p in P.values() if not np.isclose(p, 0))
+        if not np.isclose(S, 1):
+            raise ValueError("Probabilities do not sum to 1")
+        self._P: Dict[T0, float] = {x: p for x, p in P.items() if not np.isclose(p, 0)}
```
Yes, I've made this more efficient now.

ss2165 reviewed (approved):
👍
